src/main.cpp

```diff
@@ -1239,11 +1276,28 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
             return state.DoS(100, error("CheckTransaction(): txout.nValue too high"),
                              REJECT_INVALID, "bad-txns-vout-toolarge");
         nValueOut += txout.nValue;
-        if (!MoneyRange(nValueOut))
-            return state.DoS(100, error("CheckTransaction(): txout total out of range"),
-                             REJECT_INVALID, "bad-txns-txouttotal-toolarge");
     }
 
+    // Check for negative or overflow TZE output values
+    //
+    // In the case that the TZE extension is not enabled, this check
+    // is still valid, according to the false-positive semantics of this
+    // function.
+    BOOST_FOREACH(const CTzeOut& tzeout, tx.vtzeout)
+    {
+        if (tzeout.nValue < 0)
+            return state.DoS(100, error("CheckTransaction(): tzeout.nValue negative"),
+                             REJECT_INVALID, "bad-txns-vout-negative");
+        if (tzeout.nValue > MAX_MONEY)
+            return state.DoS(100, error("CheckTransaction(): tzeout.nValue too high"),
+                             REJECT_INVALID, "bad-txns-vout-toolarge");
+        nValueOut += tzeout.nValue;
+    }
+
+    if (!MoneyRange(nValueOut))
+        return state.DoS(100, error("CheckTransaction(): transparent total out of range"),
+                         REJECT_INVALID, "bad-txns-txouttotal-toolarge");
```

Comment:
Move this into the TZE output loop, to match the behaviour of the transparent output check.